File: Public/Gui/Linn/Gui/SceneGraph/Visitor.py
import sys

class DuplicateNodeNameFound(Exception):
    pass

class Visitor:
    """A Visitor used to search the scene graph."""
    def VisitNode(self, aNode):
        if aNode:
            aNode.Visit(self)
            
    def Accept(self, aNode):
        pass
    
    def AcceptNode(self, aNode):
        self.Accept(aNode)
        
    def AcceptHit(self, aHit):
        self.Accept(aHit)
        
    def AcceptPolygon(self, aPolygon):
        self.Accept(aPolygon)
        
    def AcceptText(self, aText):
        self.Accept(aText)
        
        
class SearchVisitor(Visitor):
    """Search the scene graph for a node matching aName."""
    def __init__(self, aName):
        self.iName = aName
        self.iResult = None
        
    def Search(self, aNode):
        self.VisitNode(aNode)
        return self.iResult
        
    def AcceptNode(self, aNode):
        if aNode.Name() == self.iName:
            if self.iResult:
                raise DuplicateNodeNameFound
            else:
                self.iResult = aNode
                
                
class FindGreatestZ(Visitor):
    """Seach the scene graph for the greatest z value."""
    def __init__(self):
        self.iZ = None
        
    def Search(self, aNode):
        self.VisitNode(aNode)
        return self.iZ
    
    def AcceptNode(self, aNode):
        if self.iZ:
            if aNode.WorldSrt().Translation().iZ > self.iZ:
                self.iZ = aNode.WorldSrt().Translation().iZ
        else:
            self.iZ = aNode.WorldSrt().Translation().iZ
                
                
class HitProxyVisitor(Visitor):
    """Search the scene graph for a node containing aPosition."""
    def __init__(self, aPosition):
        self.iPosition = aPosition
        self.iResults = []
        
    def Hit(self, aNode):
        self.VisitNode(aNode)
        if len(self.iResults):
            self.iResults.sort(lambda x,y: cmp(x.WorldSrt().Translation().iZ,y.WorldSrt().Translation().iZ))
            return self.iResults[-1]
        return None
    
    def AcceptHit(self, aHit):
        if aHit.Active():
            if aHit.IsInside(self.iPosition - aHit.WorldSrt().Translation()):
                self.iResults.append(aHit)


class RenderVisitor(Visitor):
    """Render the scene graph."""
    def __init__(self, aRenderer):
        self.iRenderer = aRenderer
        self.iRenderList = []
        
    def Render(self, aNode):
        self.VisitNode(aNode)
        self.iRenderList.sort(lambda x,y: cmp(x.WorldSrt().Translation().iZ,y.WorldSrt().Translation().iZ))
        for elem in self.iRenderList:
            elem.Draw(self.iRenderer)
    
    def AcceptPolygon(self, aPolygon):
        if aPolygon.Active():
            #self.iRenderer.DrawPolygon(aPolygon)
            self.iRenderList.append(aPolygon)
            
    def AcceptText(self, aText):
        if aText.Active():
            #self.iRenderer.DrawText(aText)
            self.iRenderList.append(aText)
    

class PrintVisitor(Visitor):
    """Print the scene graph structure to the console."""
    def __init__(self, aRoot):
        self.iRoot = aRoot
        
    def Print(self):
        self.VisitNode(self.iRoot)
        
    def AcceptNode(self, aNode):
        parent = aNode.Parent()
        while parent:
            parent = parent.Parent()
            print '  ',
        print aNode.Name()
                

File: Public/Gui/Linn/Gui/SceneGraph/Visitor.cs
using Linn.Gui.Scenegraph;
using System.Collections.Generic;
using System;
using Linn.Gui;

namespace Linn {
namespace Gui {
namespace Scenegraph {

public class Visitor {
    public virtual void VisitNode(Node aNode) {
        if(aNode != null) {
            aNode.Visit(this, true);
        }
    }
    
    public virtual void Accept(Node aNode) {
    }
    
    public virtual void AcceptNode(Node aNode) {
        Accept(aNode);
    }
    
    public virtual void AcceptHit(NodeHit aNode) {
        Accept(aNode);
    }
    
    public virtual void AcceptPolygon(NodePolygon aNode) {
        Accept(aNode);
    }
    
    public virtual void AcceptText(NodeText aNode) {
        Accept(aNode);
    }
    
    public virtual void AcceptList(NodeList aNode) {
        Accept(aNode);
    }
}

public class VisitorSelectProxy : Visitor
{
    public VisitorSelectProxy(Vector3d aPosition) {
        iPosition = aPosition;
    }
    
    public List<Node> Hit(Node aNode) {
        VisitNode(aNode);
        return iResults;
    }
    
    public override void AcceptHit(NodeHit aHitNode) {
        if(aHitNode.Active == true) {
            if(aHitNode.Inside(iPosition - aHitNode.WorldSrt.Translation)) {
                iResults.Add(aHitNode);
            }
        }
    }
    
    private Vector3d iPosition;
    private List<Node> iResults = new List<Node>();
}

public sealed class VisitorHitProxy : VisitorSelectProxy
{
    public VisitorHitProxy(Vector3d aPosition) : base(aPosition) {
    }
    
    public override void VisitNode(Node aNode) {
        if(aNode != null) {
            aNode.Visit(this, false);
        }
    }

    public override void AcceptHit(NodeHit aHitNode) {
        if(aHitNode.AllowHits == true) {
            base.AcceptHit(aHitNode);
        }
    }
}

public sealed class VisitorSearch : Visitor
{
    public VisitorSearch(string aName) {
        iName = aName;
    }
    
    public Node Search(Node aNode) {
        Trace.WriteLine(Trace.kGui, "VisitorSearch.Search: " + iName);
        VisitNode(aNode);
        return iResult;
    }
    
    public override void AcceptNode(Node aNode) {
        if(aNode.Fullname == iName) {
            if(iResult != null) {
                throw new DuplicateNode(aNode.Fullname);
            }
            iResult = aNode;
        }
    }
    
    private string iName;
    private Node iResult = null;
}
                
public sealed class VisitorGreatestZ : Visitor
{
    public float Search(Node aNode) {
        VisitNode(aNode);
        return iZ;
    }
    
    public override void AcceptNode(Node aNode) {
        if(iCompare == true) {
            float z = aNode.WorldSrt.Translation.Z;
            if(z > iZ) {
                iZ = z;
            }
        } else {
            iZ = aNode.WorldSrt.Translation.Z;
            iCompare = true;
        }
    }
    
    private bool iCompare = false;
    private float iZ;
}

public sealed class VisitorRender : Visitor
{
    public VisitorRender(IRenderer aRenderer) {
        iRenderer = aRenderer;
    }
    
    public override void VisitNode(Node aNode) {
        if(aNode != null) {
            aNode.Visit(this, false);
        }
    }
    
    public void Render(Node aNode) {
        Ticker render = new Ticker();
        
        Ticker visit = new Ticker(); 
        VisitNode(aNode);
        Trace.WriteLine(Trace.kRendering, "VisitorRender.Render(VisitNode): ms=" + visit.MilliSeconds);
        
        // sort the polygon list
        Ticker sort = new Ticker();
        iRenderList.Sort(new Node.NearToFar());
        Trace.WriteLine(Trace.kRendering, "VisitorRender.Render(Sort): Count=" + iRenderList.Count + ", ms=" + sort.MilliSeconds);
        
        bool success = false;
        while(!success) {
            try {
                float ms;
                iRenderer.BeginFrame();
                Ticker ticker = new Ticker();
                Trace.WriteLine(Trace.kRendering, "TickCount(b loop)=" + ticker.StartTime);
                foreach(Node node in iRenderList) {
                    Ticker drawTicker = new Ticker();
                    Trace.WriteLine(Trace.kRendering, "TickCount(b draw)=" + drawTicker.StartTime);
                    node.BeginDraw();
                    node.Draw(iRenderer);
                    node.EndDraw();
                    ms = drawTicker.MilliSeconds;
                    Trace.WriteLine(Trace.kRendering, "TickCount(e draw)=" + drawTicker.EndTime);
                    Trace.WriteLine(Trace.kRendering, "VisitorRender.Render(Draw): Fullname=" + node.Fullname + ", ms=" + ms);
                }
                ms = ticker.MilliSeconds;
                Trace.WriteLine(Trace.kRendering, "TickCount(e loop)=" + ticker.EndTime);
                Trace.WriteLine(Trace.kRendering, "VisitorRender.Render: loop ms=" + ms);
                iRenderer.EndFrame();
                success = true;
            } catch(System.InvalidOperationException e) {
                System.Console.WriteLine("VisitorRender.Render:\n" + e);
            }
        }
        
        iRenderer.Stats.UpdateTotalMs(render.MilliSeconds);
        Trace.WriteLine(Trace.kRendering, "VisitorRender.Render: ms=" + render.MilliSeconds);
    }
    
    public override void AcceptPolygon(NodePolygon aPolygon) {
        if(aPolygon.Active) {
            iRenderList.Add(aPolygon);
        }
    }
    
    public override void AcceptText(NodeText aText) {
        if(aText.Active) {
            iRenderList.Add(aText);
        }
    }
    
    public override void AcceptList(NodeList aList) {
        if(aList.Active) {
            //aList.Update();
            iRenderList.Add(aList);
        }
    }
    
    private IRenderer iRenderer;
    private List<Node> iRenderList = new List<Node>();
}

public sealed class VisitorPrint : Visitor
{
    public void Print(Node aNode) {
        VisitNode(aNode);
    }
    
    public void NodePrint(Node aNode) {
        Node parent = aNode.Parent;
        while(parent != null) {
            parent = parent.Parent;
            Console.Write("  ");
        }
        Console.WriteLine(aNode.Fullname);
    }
    
    public override void AcceptNode(Node aNode) {
        NodePrint(aNode);
    }
}

} // Scenegraph
} // Gui
} // Linn
